xarray/core/dataset.py

```diff
@@ -1210,10 +1210,20 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         needed_dims = set(variable.dims)
 
         coords: dict[Hashable, Variable] = {}
+        temp_indexes = self.xindexes
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
-                coords[k] = self._variables[k]
+            if k in self._coord_names:
+                if (
+                    k not in coords
+                    and k in temp_indexes
+                    and set(temp_indexes.get_all_dims(k)) & needed_dims
+                ):
+                    # add all coordinates of each index that shares at least one dimension
+                    # with the dimensions of the extracted variable
+                    coords.update(temp_indexes.get_all_coords(k))
```

Comment:
Coud this use a separate loop after the existing loop instead? e.g.,
```python
for k in self._indexes:
    if k in coords:
        coords.update(self.xindexes.get_all_coords(k))
```

Or if we allow indexes without a coordinate of the same name:
```python
for k in self._indexes:
    if set(self.xindexes.get_all_dims(k)) & needed_dims:
        coords.update(self.xindexes.get_all_coords(k))
```

Ideally, I would like the logic here to be just as simple as the words describing how it works, so a comment is not necessary!